### PR TITLE
[WIP] uniq command

### PR DESCRIPTION
**Pull request by bertinatto**
Hi there,

Would you mind taking a look at what I've so far for the `uniq` command? 

I started off using the code from `sort`. The code from `repeated_lines` and `unique_lines` are based in the code from `std::vec:Vec::dedup_by`.

I'd love to get some feedback from more experienced Rust developers! 😃 

Thanks!

### CODE REVIEW

**Comment by simondesloges** [comment]
It seems the functions `repeated_lines` and `unique_lines` don't always return the right thing. [I tested them a bit and they broke easily](https://play.rust-lang.org/?gist=ef477de53f3ed65a152f1e7f57ee2867&version=nightly&backtrace=0).

`unique_lines` has trouble when the repeated lines are in odd numbers (3, 5, 7...).
`repeated_lines` seems to always keep the last line, no matter what, and also has trouble when duplicates are repeated too many times.

I would recommend implementing a function that would "squash" repeated lines of a vector and then store a tuple of the number of repetitions and the string itself in another vector. This would make it trivial to implement `repeated_lines` and `unique_lines`, would allow code to be reused by both and would pave the way to add the -c/--count flag.

pseudo code exemple:
```rust
let v = the_squash_func(["hello", "hello", "hello", "hi", "hoy"]);
assert!(v == [(3, "hello"), (1, "hi"), (1, "hoy")]);
```

Everything else seemed fine.

**Comment by bertinatto** [comment]
Thanks for your very nice suggestions, @simondesloges. Very appreciated.

**Comment by simondesloges** [comment]
Nice. I tested `get_squashed_lines`, `unique_lines` and `repeated_lines`, and it seemed they worked fine.

Some very minor nitpicks: `ArgParser::new(2)` should probably be `ArgParser::new(5)` (because uniq has 5 flags), I would have personally chosen the name `ignore_case` instead of `icase` and function `eq_strings` could take arguments as `&str` instead of `&String`.

In this current state, I would be ok with this being merged.

But I don't believe this is perfect, though.
Let's look at `get_squashed_lines`: it takes ownership of the `lines` argument (the vector is moved rather than borrowed), then `.clone()` the strings within that vector and returns a new vector with those cloned strings in it. Certainly, this `.clone()` is superfluous, I am sure the strings could be moved. Now, the real problem is to rewrite the algorithm in a way that satisfy the borrow checker.
The only reason I bother talking about this is because cloning stuff to circumvent the borrow checker is seen as an anti-pattern. Other than that, the only downside to using `.clone()` is performance related (I think), and I don't think coreutils need to bother with performance right now. Working properly and proper error handling/reporting should be our goal for now.

File: src/bin/uniq.rs [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+    }
+    Ok(lines)
+}
+
+fn lines_from_files(paths: &Vec<&String>) -> Result<Vec<String>, Error> {
+    let mut lines = Vec::new();
+
+    for path in paths {
+        let f = try!(File::open(path));
+        let f = BufReader::new(f);
```
Comment (mmstick) [inline src/bin/uniq.rs]: You could write:

```rust
let file = Bufreader::new(File::open(path)?);
```

File: src/bin/uniq.rs [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+        for line in f.lines() {
+            let line = match line {
+                Ok(l) => l,
+                Err(e) => return Err(e),
+            };
+            lines.push(line);
```
Comment (mmstick) [inline src/bin/uniq.rs]: This could be shortened to:

```rust
lines.push(line?);
```

File: src/bin/uniq.rs [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+        exit(0);
+    }
+
+    let lines = match parser.args.is_empty() {
+        true => lines_from_stdin(),
+        false => {
+            let mut paths = Vec::new();
+            for dir in parser.args.iter() {
+                paths.push(dir);
+            }
```
Comment (mmstick) [inline src/bin/uniq.rs]: You could shorten this to:

```rust
let paths = parser.args.iter().collect::<Vec<_>>();
```

File: src/bin/uniq.rs [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+        let f = BufReader::new(f);
+        for line in f.lines() {
+            let line = match line {
+                Ok(l) => l,
+                Err(e) => return Err(e),
+            };
+            lines.push(line);
+        }
+    }
+    Ok(lines)
```
Comment (mmstick) [inline src/bin/uniq.rs]: You could even just write the whole thing here as:

```rust
Ok(paths.iter().flat_map(|path| {
    Bufreader::new(File::open(path)?).lines().map(|x| line?)
}).collect::<Vec<String>>())
```

Comment (bertinatto) [inline src/bin/uniq.rs]: Thanks, @mmstick. I learned a lot from your suggestions.

I wasn't able to implement this one, though. Apparently, the `?` doesn't work inside a closure and I had to use `unwrap()` instead, which seemed inappropriate.

Perhaps am I missing something?

**Comment by bertinatto** [comment]
@simondesloges, thanks for your valuable feedback.

> cloning stuff to circumvent the borrow checker is seen as an anti-pattern

Hm, interesting.

The closest I could get without `clone`'ing or going down the `usafe` road was using `mem::replace`. Still it doesn't seem to be the ideal, thought. The main problem I'm facing is that every time I move an element from `lines` the entire vector is taken over, which makes the comparison with the next element cumbersome.

Even if either the `clone` or the `mem::replace` approaches are alright in this case, I'd love to know if there's a better pattern to solve this issue. I appreciate any other suggestions that you might have.

File: src/bin/uniq.rs [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+
+            if parser.found("count") {
+                for (c, v) in squashed {
+                    println!("{} {}", c, v);
+                }
+            } else {
+                for (_, v) in squashed {
+                    println!("{}", v);
+                }
+            }
```
Comment (mmstick) [inline src/bin/uniq.rs]: It should be more efficient to buffer 8K worth of bytes at a time before printing to standard output, to avoid having an excessive amount of syscalls when there are a lot of lines being processed.

File: src/bin/uniq.rs [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+        true => lines_from_stdin(),
+        false => {
+            let paths = parser.args.iter().collect::<Vec<_>>();
+            lines_from_files(&paths)
+        }
+    };
+
+    match lines {
+        Ok(mut l) => {
+            let mut squashed = get_squashed_lines(&mut l, parser.found("ignore-case"));
```
Comment (mmstick) [inline src/bin/uniq.rs]: Does `l` need to be mutable if you have no plans to use it later?

**Comment by mmstick** [comment]
> The closest I could get without clone'ing or going down the usafe road was using mem::replace. Still it doesn't seem to be the ideal, thought. The main problem I'm facing is that every time I move an element from lines the entire vector is taken over, which makes the comparison with the next element cumbersome.

Even if either the clone or the mem::replace approaches are alright in this case, I'd love to know if there's a better pattern to solve this issue. I appreciate any other suggestions that you might have.

@bertinatto From what I can see, you need to change the signature of your function like so:

```rust
fn get_squashed_lines<'a>(lines: &'a Vec<String>, ignore_case: bool) -> Vec<(usize, &'a String)>
```

Instead of mutating the lines vector, assign a lifetime to it and have your new vector point to references in that string, rather than creating copies of each string.

File: src/bin/uniq.rs [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+    --unique-lines
+        only print unique lines
+"#; /* @MANEND */
+
+fn lines_from_stdin() -> Result<Vec<String>, Error> {
+    let stdin = stdin();
+    let mut lines = Vec::new();
+
+    let f = BufReader::new(stdin.lock());
+    for line in f.lines() {
```
Comment (mmstick) [inline src/bin/uniq.rs]: From a performance perspective, you don't need the String type or the lines method at all in your program. You can change String to Vec<u8> and have the iterator split on occurrences of `b\n` instead, so that you don't have to pay for UTF8 validation.

Comment (bertinatto) [inline src/bin/uniq.rs]: That's interesting. I copied this part of the code came from `sort`. 

I was able to convert most of the program to use `Vec<u8>` instead of `String`. The only remaining part is the case-insensitive comparison of two `Vec<u8>`. However, I looked at other utilities and noticed that they also use `lines()` and `String`, and I wondered if their performance disadvantages would be too expressive for `unique`. 

Do you think I can keep using them for now or do you think the performance disadvantages of using `String` and `lines()` are too high?

Comment (mmstick) [inline src/bin/uniq.rs]: The other commands should me modified to remove their reliance upon `String` and it's UTF-8 validation. As for performing a case-sensitive comparison, you can check if each u8 is between the range of 65 ('A') and 90 ('Z'), and adding 32 to that number will get the lowercase character's value.

Comment (bertinatto) [inline src/bin/uniq.rs]: Sweet! Thanks!

Comment (simondesloges) [inline src/bin/uniq.rs]: Regarding the case-insensitive comparison, are these optimisations still valid if the input is not pure ascii? It would only compare ascii characters in a case-insensitive manner, but other characters would not be compared properly, like é and É. If the goal is to match, say, GNU uniq's behavior, then this is ok. If (for some reasons) we want to properly support unicode, it's not valid.

Comment (mmstick) [inline src/bin/uniq.rs]: Only ascii characters would be checked for case-insensitiveness.

**Comment by mmstick** [comment]
I think this can be merged as is right now, if @bertinatto wants to. Further optimizations can come later.

**Comment by bertinatto** [comment]
@mmstick, no worries; it's good for my learning.

File: src/bin/uniq.rs (hunk 5 above)
Comment (bertinatto) [inline src/bin/uniq.rs]: Can you take a look if my changes reflect what you meant?

File: src/bin/uniq.rs (hunk 7 above)
Comment (bertinatto) [inline src/bin/uniq.rs]: It looks like it works:

```
echo "CAFÉ\nCafé\nCoração\nCORAÇÃO"  | target/debug/uniq  -ci

2 CAFÉ
2 Coração
```

**Comment by jackpot51** [comment]
Looks really good to me! I have not used BufWriter, but I really need to!

**Comment by mmstick** [comment]
I'll also add that you can write your code like this to avoid all the unwrapping:

```rust
let _ = stdout.write_all(c)
    .and_then(|_| stdout.write_all(b" "))
    .and_then(|_| stdout.write_all(v))
    .and_then(|_| stdout.write_all(b"\n"));
```
